test/connector/tcp/mssql/test

```diff
@@ -1,17 +1,133 @@
-#!/bin/bash
+#!/bin/bash -e
 
-export SECRETLESS_HOST=secretless
+readonly db_host_tls="mssql"
+readonly db_host_no_tls="mssql" # TODO: configure a non-ssl container?
+readonly db_port=1433
+readonly db_user="sa"
+readonly db_password="yourStrong()Password"
+readonly secretless_port=2223
+secretless_host="secretless"
+benchmark=false
+
+help_and_exit() {
+  local retval=${1:-1}
+  cat <<EOF
+This script tests access to an MSSQL server, first accessing the
+server directly (to establish a testing baseline), and then
+via a secretless broker.
+
+This script detects if you're in "dev" mode by the existence of
+the secretless-dev container.  That container will exist only if
+the development process was started using the "./dev" command.
+
+Options:
+  -b runs benchmarks
+EOF
+  exit "$retval"
+}
+
+while getopts :b opt; do
+    case $opt in
+        b) benchmark=true;;
+       \?) echo "Unknown option -$OPTARG"; help_and_exit 1;;
+    esac
+done
+
+assert_contains() {
+  # assert_contains <string> <substring>
+  local string=$1
+  local substring=$2
+
+  if [[ "${string}" =~ ${substring} ]]; then
+      echo "."
+      return
+  fi
+
+  echo ""
+  echo "FAIL:"
+  echo ""
+  echo "The string:"
+  echo ""
+  echo "    ${string}"
+  echo ""
+  echo "Did not contain expected substring:"
+  echo ""
+  echo "    ${substring}"
+  echo ""
+
+  exit 1
+}
+
+query_sql_service() {
+  # query_sql_service <server> <port> <user> <password>
+  server=$1
+  port=$2
+  user=$3
+  pw=$4
+
+  docker_cmd="docker-compose run --rm --no-deps"
+  sql_cmd="sqlcmd -S \"$server,$port\" -U \"$user\" -P \"$pw\" -Q \"SELECT 1+1\""
```

Comment:
Fly by comment:

First, this is a very clean bash.  A trick to avoid all the quote escaping when you have this kind of double substitution situation -- creating a variable (that itself has expanded variables) that will be expanded later-- is to use bash arrays.  If you do:
```sh
sql_cmd=(sqlcmd -S "$server,$port" -U "$user" -P "$pw" -Q "SELECT 1+1")
```
and then later when you use it (note the required quotes around the array which expands it with each element quoted):
```sh
$docker_cmd test bash -c "${sql_cmd[@]}" 2>&1 || true
```
everything will be kosher.   A small improvement imo.

Reply:
Ah, yes, thanks! I had tried that earlier using bash arrays, but it wasn't working because I was passing SQL command via `bash -c`. Removing the `bash -c` then allows the bash array trick to work.
Thanks for the explanation `note the required quotes around the array which expands it with each element quoted`, that was the piece of the puzzle that I was missing to understand how this bash array expansion works!